### PR TITLE
Close #12331: Use CoordsXY in track_remove_station_element

### LINKED ISSUE #12331: Use CoordsXY in track_remove_station_element
Refactor `track_remove_station_element` to take a `const CoordsXYZD& ` instead of the individual components. Take care with the Z value as it is currently in the form of smallZ and CoordsXYZD take bigZ (`*COORDS_Z_STEP`). There are a number of loops that should be refactored to also use `CoordsXY` structures instead of `removeX/Y, stationX/Y0/1`. When doing this refactor do not do all of the changes in one commit. Split it up into a number of commits. One for changing the function signature and callers, one commit for changing `removeX/Y` so on and so on.

Please remember the commit style: https://github.com/OpenRCT2/OpenRCT2/wiki/Commit-Messages
And the coding style: https://github.com/OpenRCT2/OpenRCT2/wiki/Coding-Style

### PR DESCRIPTION
**Pull request by ion232**
Closes #12331: Use CoordsXY in track_remove_station_element. The function signature of track_remove_station_element, as well 
as some local variables were refactored to use appropriate Coords structs.

### CODE REVIEW

File: src/openrct2/actions/TrackRemoveAction.hpp [hunk 1]
```diff
@@ -216,7 +216,7 @@ DEFINE_GAME_ACTION(TrackRemoveAction, GAME_COMMAND_REMOVE_TRACK, GameActionResul
 
             if (entranceDirections & TRACK_SEQUENCE_FLAG_ORIGIN && (tileElement->AsTrack()->GetSequenceIndex() == 0))
             {
-                if (!track_remove_station_element(mapLoc.x, mapLoc.y, mapLoc.z / 8, _origin.direction, rideIndex, 0))
+                if (!track_remove_station_element({ mapLoc.x, mapLoc.y, mapLoc.z / 8, _origin.direction }, rideIndex, 0))
```
Comment (duncanspumpkin) [inline src/openrct2/actions/TrackRemoveAction.hpp:219]: CoordsXYZ should never have a /8 on its Z component. You need to go through and remove the multiplies and divides of the z component. The /8 means that this was a smallZ function we want the whole codebase to be in bigZ

Comment (ion232) [reply src/openrct2/actions/TrackRemoveAction.hpp:219]: @duncanspumpkin Will do, thanks for the clarification on smallZ vs bigZ

File: src/openrct2/ride/Track.cpp [hunk 2]
```diff
@@ -828,54 +826,55 @@ bool track_remove_station_element(int32_t x, int32_t y, int32_t z, Direction dir
...
 
-        stationX0 = x;
-        stationY0 = y;
+        station0.x = currentLoc.x;
+        station0.y = currentLoc.y;
         byte_F441D1++;
 
-        x -= CoordsDirectionDelta[direction].x;
-        y -= CoordsDirectionDelta[direction].y;
+        currentLoc.x -= CoordsDirectionDelta[loc.direction].x;
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:846]: We have overloads for minus. So you can do currentLoc -= CoordsDd[x];

File: src/openrct2/ride/Track.cpp [hunk 3]
```diff
@@ -828,54 +826,55 @@ bool track_remove_station_element(int32_t x, int32_t y, int32_t z, Direction dir
...
-    x = stationX1;
-    y = stationY1;
+    currentLoc.x = station1.x;
+    currentLoc.y = station1.y;
     do
     {
-        x += CoordsDirectionDelta[direction].x;
-        y += CoordsDirectionDelta[direction].y;
+        currentLoc.x += CoordsDirectionDelta[loc.direction].x;
+        currentLoc.y += CoordsDirectionDelta[loc.direction].y;
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:856]: Use the overloads

File: src/openrct2/ride/Track.cpp [hunk 4]
```diff
@@ -828,54 +826,55 @@ bool track_remove_station_element(int32_t x, int32_t y, int32_t z, Direction dir
...
+    while (
+        (stationElement = find_station_element({ currentLoc.x, currentLoc.y, loc.z * COORDS_Z_STEP, loc.direction }, rideIndex))
+        != nullptr)
     {
         if (stationElement->AsTrack()->GetTrackType() == TRACK_ELEM_END_STATION)
         {
             if (flags & GAME_COMMAND_FLAG_APPLY)
             {
-                ride_remove_station(ride, { x, y, z * COORDS_Z_STEP });
+                ride_remove_station(ride, { currentLoc.x, currentLoc.y, loc.z * COORDS_Z_STEP });
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:838]: Z

File: src/openrct2/ride/Track.cpp [hunk 5]
```diff
@@ -828,54 +826,55 @@ bool track_remove_station_element(int32_t x, int32_t y, int32_t z, Direction dir
...
+    currentLoc.y = station1.y;
     do
     {
-        x += CoordsDirectionDelta[direction].x;
-        y += CoordsDirectionDelta[direction].y;
+        currentLoc.x += CoordsDirectionDelta[loc.direction].x;
+        currentLoc.y += CoordsDirectionDelta[loc.direction].y;
 
-        stationElement = find_station_element({ x, y, z * COORDS_Z_STEP, direction }, rideIndex);
+        stationElement = find_station_element({ currentLoc.x, currentLoc.y, loc.z * COORDS_Z_STEP, loc.direction }, rideIndex);
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:858]: Z

File: src/openrct2/ride/Track.cpp [hunk 6]
```diff
@@ -828,54 +826,55 @@ bool track_remove_station_element(int32_t x, int32_t y, int32_t z, Direction dir
...
+            station1.x = currentLoc.x;
+            station1.y = currentLoc.y;
             stationLength++;
         }
     } while (stationElement != nullptr);
 
     if (!(flags & GAME_COMMAND_FLAG_APPLY))
     {
-        if ((removeX != stationX0 || removeY != stationY0) && (removeX != stationX1 || removeY != stationY1)
+        if ((remove.x != station0.x || remove.y != station0.y) && (remove.x != station1.x || remove.y != station1.y)
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:877]: Use the overloaded comparison operators

File: src/openrct2/ride/Track.cpp [hunk 7]
```diff
@@ -887,21 +886,23 @@ bool track_remove_station_element(int32_t x, int32_t y, int32_t z, Direction dir
...
-    y = stationY1;
+    currentLoc.x = station1.x;
+    currentLoc.y = station1.y;
     bool finaliseStationDone;
     do
     {
         finaliseStationDone = true;
 
-        if (x != removeX || y != removeY)
+        if (currentLoc.x != remove.x || currentLoc.y != remove.y)
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:896]: Use overloaded comparison

File: src/openrct2/ride/Track.cpp [hunk 8]
```diff
@@ -887,21 +886,23 @@ bool track_remove_station_element(int32_t x, int32_t y, int32_t z, Direction dir
...
         {
-            stationElement = find_station_element({ x, y, z * COORDS_Z_STEP, direction }, rideIndex);
+            stationElement = find_station_element(
+                { currentLoc.x, currentLoc.y, loc.z * COORDS_Z_STEP, loc.direction }, rideIndex);
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:899]: Z

File: src/openrct2/ride/Track.cpp [hunk 9]
```diff
@@ -910,9 +911,9 @@ bool track_remove_station_element(int32_t x, int32_t y, int32_t z, Direction dir
                     }
                     else
                     {
-                        ride->stations[stationIndex].Start.x = x;
-                        ride->stations[stationIndex].Start.y = y;
-                        ride->stations[stationIndex].Height = z;
+                        ride->stations[stationIndex].Start.x = currentLoc.x;
+                        ride->stations[stationIndex].Start.y = currentLoc.y;
+                        ride->stations[stationIndex].Height = loc.z;
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:916]: Possible bug in existing code. Need to investigate. Unsure why there is no multiply.

File: src/openrct2/ride/Track.cpp [hunk 10]
```diff
@@ -887,21 +886,23 @@ bool track_remove_station_element(int32_t x, int32_t y, int32_t z, Direction dir
...
-            stationElement = find_station_element({ x, y, z * COORDS_Z_STEP, direction }, rideIndex);
+            stationElement = find_station_element(
+                { currentLoc.x, currentLoc.y, loc.z * COORDS_Z_STEP, loc.direction }, rideIndex);
             if (stationElement != nullptr)
             {
                 int32_t targetTrackType;
-                if ((x == stationX1 && y == stationY1)
-                    || (x + CoordsDirectionDelta[direction].x == removeX && y + CoordsDirectionDelta[direction].y == removeY))
+                if ((currentLoc.x == station1.x && currentLoc.y == station1.y)
+                    || (currentLoc.x + CoordsDirectionDelta[loc.direction].x == remove.x
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:904]: Use the overloads

File: src/openrct2/ride/Track.cpp [hunk 11]
```diff
@@ -923,13 +924,14 @@ bool track_remove_station_element(int32_t x, int32_t y, int32_t z, Direction dir
                 }
                 else
                 {
-                    if (x - CoordsDirectionDelta[direction].x == removeX && y - CoordsDirectionDelta[direction].y == removeY)
+                    if (currentLoc.x - CoordsDirectionDelta[loc.direction].x == remove.x
+                        && currentLoc.y - CoordsDirectionDelta[loc.direction].y == remove.y)
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:928]: Overloads

File: src/openrct2/ride/Track.cpp [hunk 12]
```diff
@@ -923,13 +924,14 @@ bool track_remove_station_element(int32_t x, int32_t y, int32_t z, Direction dir
...
                     {
                         targetTrackType = TRACK_ELEM_BEGIN_STATION;
                     }
                     else
                     {
-                        if (x == stationX0 && y == stationY0)
+                        if (currentLoc.x == station0.x && currentLoc.y == station0.y)
                         {
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:935]: Overloads

File: src/openrct2/ride/Track.cpp [hunk 13]
```diff
@@ -941,14 +943,14 @@ bool track_remove_station_element(int32_t x, int32_t y, int32_t z, Direction dir
...
-                map_invalidate_element({ x, y }, stationElement);
+                map_invalidate_element({ currentLoc.x, currentLoc.y }, stationElement);
             }
         }
 
-        if (x != stationX0 || y != stationY0)
+        if (currentLoc.x != station0.x || currentLoc.y != station0.y)
         {
-            x -= CoordsDirectionDelta[direction].x;
-            y -= CoordsDirectionDelta[direction].y;
```
Comment (duncanspumpkin) [inline src/openrct2/ride/Track.cpp:951]: Overloads

File: src/openrct2/actions/TrackRemoveAction.hpp [hunk 14]
```diff
@@ -406,7 +406,7 @@ DEFINE_GAME_ACTION(TrackRemoveAction, GAME_COMMAND_REMOVE_TRACK, GameActionResul
 
             if (entranceDirections & TRACK_SEQUENCE_FLAG_ORIGIN && (tileElement->AsTrack()->GetSequenceIndex() == 0))
             {
-                if (!track_remove_station_element(mapLoc.x, mapLoc.y, mapLoc.z / 8, _origin.direction, rideIndex, 0))
+                if (!track_remove_station_element({ mapLoc.x, mapLoc.y, mapLoc.z, _origin.direction }, rideIndex, 0))
```
Comment (duncanspumpkin) [inline src/openrct2/actions/TrackRemoveAction.hpp:409]: One last issue you can still use the other constructors to reduce having to get the components individually. Same applies to the other callers

**Comment by ion232** [comment]
> 
> 
> All looks great now. I'm still wondering about the stations[stationIndex].Height part. I need to check to see if that is correct. It's quite possible that this is an existing bug.

I would assume it's a bug, given the '/ COORDS_Z_STEP' that occurs in track_add_station_element in a similar context.

**Comment by duncanspumpkin** [comment]
Looks like Height is in terms of smallZ so this is all good.

**Pull request merged**